jni/events/sampleevent.cpp

```diff
@@ -182,17 +176,197 @@ void SampleEvent::setSample( AudioBuffer* sampleBuffer )
+        // use BaseAudioEvent behaviour if no custom playback rate is requested
+        BaseAudioEvent::mixBuffer( outputBuffer, bufferPosition, minBufferPosition, maxBufferPosition,
                                    loopStarted, loopOffset, useChannelRange );
+        return;
+    }
+
+    // custom playback rate
+
+    int bufferSize = outputBuffer->bufferSize;
+
+    // if the output channel amount differs from this events channel amount, we might
+    // potentially have a bad time (e.g. engine has mono output while this event is stereo)
+    // ideally events should never hold more channels than AudioEngineProps::OUTPUT_CHANNELS
+
+    int outputChannels = std::min( _buffer->amountOfChannels, outputBuffer->amountOfChannels );
+    int i, t, c, ca;
+    float frac;
+    SAMPLE_TYPE* srcBuffer;
+    SAMPLE_TYPE* tgtBuffer;
+    SAMPLE_TYPE s1, s2;
+
+    // at custom playback rate we require floating point precision
+    float eventStart = ( float ) _eventStart;
+    float eventEnd   = ( float ) getEventEnd();
+    float mbp = ( float ) maxBufferPosition;
+    float lo  = ( float ) loopOffset;
+    // iterator that increments by the playback rate
+    float fi = 0.f;
+
+    // take sequencer playhead position and determine what the position
+    // should be relative to this events playback rate
+    float readPointerStart = eventStart + ((( float ) bufferPosition - eventStart ) * _playbackRate );
+    float readPointer;
+
+    // non-loopeable event whose playback is tied to the Sequencer
+
+    if ( !_loopeable )
+    {
+        for ( i = 0; i < bufferSize; ++i, fi += _playbackRate )
+        {
+            // read pointer progresses by the playback rate
+            readPointer = readPointerStart + fi;
+
+            // over the max position ? read from the start ( implies that sequence has started loop )
+            if ( readPointer > mbp )
+            {
+                if ( useChannelRange )
+                    readPointer -= mbp;
+
+                else if ( !loopStarted )
+                    break;
+            }
+
+            if ( readPointer >= eventStart && readPointer <= eventEnd )
+            {
+                // mind the offset ! ( source buffer starts at 0 while
+                // the _eventStart defines where the event is positioned
+                // subtract it from current sequencer pointer to get the
+                // offset relative to the source buffer
+                readPointer -= eventStart;
+
+                t    = ( int ) readPointer;
+                frac = readPointer - t; // between 0 - 1 range
+
+                for ( c = 0; c < outputChannels; ++c )
+                {
+                    srcBuffer = _buffer->getBufferForChannel( c );
+                    tgtBuffer = outputBuffer->getBufferForChannel( c );
+
+                    s1 = srcBuffer[ t ];
+                    s2 = srcBuffer[ t + 1 ];
+
+                    tgtBuffer[ i ] += (( s1 + ( s2 - s1 ) * frac ) * _volume );
+                }
```

Comment:
Linear interpolation of the sample by using fractions to determine the "playback head" iterator position.